### ISSUE TITLE
Wrong assignment between Array<TinyVector> and TinyVector expression

### ISSUE THREAD

**Issue by AnderOne** [blitzpp/blitz#108]
There is a problem with to unwrap of expression templates when we use multicomponent Array on the left side and a vector expression on the right.
```C++
#include <blitz/array.h>
#include <iostream>

int main() {

	blitz::Array<blitz::TinyVector<double, 2>, 1> DAT(5);
	blitz::TinyVector<double, 2> TMP(3, 7);
#if false
	//Here we construct TinyVector from expression and fill Array. It's right!
	DAT = blitz::TinyVector<double, 2> (TMP * 2 - 1);
#else
	//This code leads to incorrect behavior!
	DAT = TMP * 2 - 1;
#endif
	std::cout << DAT << std::endl;

	return 0;
}
```

If N-dimensional array is used then it leads to compilation error:
```C++
#include <blitz/array.h>
#include <iostream>

int main() {

	blitz::Array<blitz::TinyVector<double, 2>, 2> DAT(5, 4);
	blitz::TinyVector<double, 2> TMP(3, 7);
#if false
	//Here we construct TinyVector from expression and fill Array. It's right!
	DAT = blitz::TinyVector<double, 2> (TMP * 2 - 1);
#else
	//This code leads to compilation error!
	/*./blitz/blitz/array/expr.h:194:81:
	error: no match for call to
	‘(const T_expr {aka const blitz::FastTV2Iterator<double, 2>})
	(const blitz::TinyVector<int, 2>&)’
	T_result operator()(const TinyVector<int, N_rank>& i) const
	{ return iter_(i); }
	*/
	DAT = TMP * 2 - 1;
#endif
	std::cout << DAT << std::endl;

	return 0;
}
```

**Comment by citibeth** [comment]
I had no idea that tinyvector and array ever even mixed like this.  Can we
get rid of tinyvector in favor of std::array?  I think that would be an
improvement

**Comment by AnderOne** [comment]
I could suggest make TinyVector as scalar type without expression template support.
Then all arithmetic operators between TinyVectors will return TinyVector as value.
Also assignment operator can get appropriate specialization.
It's simplest way to solve this problem.

**Comment by lutorm** [comment]
TinyVector supports templated, vectorized expressions. Its performance is
crucial for heavy use of fixed-size vectors (e.g. 3-d vectors etc).
std::array does not support any of that.

While this behavior for multicomponent arrays clearly is a bug, if you
don't want the expression template functionality for TinyVector it's
trivial to write your own vector class and use that instead.

**Comment by AnderOne** [comment]
It seems that I have solved the problem.
But for this purpose I had to use the means of C++11 (from <type_traits>):
https://github.com/AnderOne/blitz/commits/test_for_expr

I'm not sure that my code works correctly in all possible cases,
but it passes all existing tests.

If you decide that it's useful, I could send a pull request.

**Comment by citibeth** [comment]
A PR would be great!

**Comment by AnderOne** [comment]
PR is here: #109

**Comment by lutorm** [comment]
Sorry for the slow response, I was traveling.

TinyVector is used for the internal offset and bounds calculations in
blitz, and for the blocked Array expressions used to facilitate SIMD
optimization. *But* it's also a user-facing class that can be used for any
fixed-size vector calculations where high performance is needed. In my
physics codes I use TinyVector/TinyMatrix extensively for 3d vector
calculations. std::array does not support math operators. Even if it did,
it would not have the performance of TinyVector. (Fixed-size
expression-template vector/matrix classes that don't allocate any memory
dynamically are much more efficient than Array for short lengths, so it is
not a substitute either.)

Cheers,

/Patrik

**Comment by citibeth** [comment]
I believe std::array is a fixed size template that doesn’t allocate memory.
???

**Comment by lutorm** [comment]
This is true, but it has no math operators, let alone ones using something
similar to the blitz ET machinery.

**Comment by AnderOne** [comment]
My patch doesn't break ET logic. It just make checking for a type conversion between
right side expression and Array::T_numtype in a compile time.
However, this feature requires C++11.

I known that migrate to a modern C++ repeatedly discussed.
But in this case, it really
fixes a bug.

**Comment by citibeth** [comment]
On Sun, Apr 7, 2019 at 7:15 PM Patrik Jonsson <notifications@github.com>
wrote:

> This is true, but it has no math operators, let alone ones using something
> similar to the blitz ET machinery.
>

In my experience, the non-standard TinyVector class requires I frequently
copy stuff from std::array in to a TinyVector for dimension sizes, indices,
etc.  That's the main impetus for moving to std::array.  Math operators can
be defined on std::array if one wishes.  Moreover, I'm wondering if
features already provided by Eigen might be a better fit for the
more-extensive application of TinyVector described here.
